src/lxml/sax.py

```diff
@@ -235,37 +236,31 @@ def _recursive_saxify(self, element):
         if element.text:
             content_handler.characters(element.text)
         for child in element:
-            self._recursive_saxify(child)
+            self._recursive_saxify(child, element_nsmap)
         content_handler.endElementNS((ns_uri, local_name), qname)
         for prefix, uri in new_prefixes:
             content_handler.endPrefixMapping(prefix)
         if element.tail:
             content_handler.characters(element.tail)
 
-    def _build_qname(self, ns_uri, local_name, prefixes, preferred_prefix):
+    def _build_qname(self, ns_uri, local_name, nsmap, preferred_prefix,
+                     is_attribute):
         if ns_uri is None:
             return local_name
 
-        if prefixes.get(preferred_prefix) == ns_uri:
+        if nsmap.get(preferred_prefix) == ns_uri and not is_attribute:
             prefix = preferred_prefix
         else:
             # Pick the first matching prefix:
-            for pfx in sorted(prefixes, key=str):
-                if prefixes[pfx] == ns_uri:
-                    prefix = pfx
-                    if pfx is None and preferred_prefix == -1:
-                        # If preferred_prefix is -1, that's a flag to say
-                        # that we want a prefix, any prefix, and only
-                        # accept the default prefix if no other is
-                        # available
-                        continue
-                    break
+            prefix = min(pfx for (pfx, uri) in nsmap.items()
+                         if pfx is not None and uri == ns_uri)
```

Comment:
One more thing to think about: this will fail now if there is only the default prefix or if the URI does not have a prefix. (`min()` requires at least one value.) The second case probably shouldn't happen (and could be considered an actual error case), but the first seems ok and shouldn't fail.